BugFix: make date display of profile's history follow selected UI locale

SlySven opened the pull request:
This PR adds a `QLocale` object to the `mudlet` class that is set to the user's selection of UI locale (language + country) so that items that need to return data in a locale aware form can use: `(const QLocale&) mudlet::getUserLocale()` in order to present their information in their selected form rather than using what they or their OS/DE has set system/desktop-wide.

In this PR it is used to format and present in the correct language the date and time entries for the profile's history but it can be used elsewhere within the Mudlet application as needed.

Also:
* reorder some items in `dlgConnectionProfile`'s class to be correct.
* mark an unused argument in `dlgConnectionProfile::slot_update_name(...)` as unused to prevent a warning.

This PR came about as part of a discussion in the Mudlet Discord Guild in the (restricted to *Translator* role and above) **\#translation** channel around 2020/06/17-18...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: make date display of profile's history follow selected UI locale

This PR adds a `QLocale` object to the `mudlet` class that is set to the
user's selection of UI locale (language + country) so that items that need
to return data in a locale aware form can use:
`(const QLocale&) mudlet::getUserLocale()`
in order to present their information in their selected form rather than
using what they or their OS/DE has set system/desktop-wide.

In this PR it is used to format and present in the correct language the
date and time entries for the profile's history but it can be used
elsewhere within the Mudlet application as needed.

Also:
* reorder some items in `dlgConnectionProfile`'s class to be correct.
* mark an unused argument in `dlgConnectionProfile::slot_update_name(...)`
as unused to prevent a warning.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
>  a discussion in the Mudlet Discord Guild in the **#translation** channel around 2020/06/17-18...

Background is Mudlet was showing information in 2 different locales simultaneously: 1) Locale selected in OS settings and 2) Locale selected in Mudlet preferences. With this PR going forward, only setting # 2 should be relevant for all data shown.

![](https://i.imgur.com/2ihhplP.png)

Kebap commented:
Looking at this test build, it seems mostly fine BUT shows a new unneccessary data:

![image](https://user-images.githubusercontent.com/117238/85064786-60845d00-b1ac-11ea-952d-0d2726cc6635.png)

There was unexpected text appended in the end, which represents the abbreviation MEST (middle European summer time)
This destroys layout and should go away again. There was no time zone indication before this PR.

SlySven committed:
Revise: exclude the display of a time zone in the profile's history

It takes up extra space which is tight in that widget and (apart from the
time around a DST change) does not impart much information to the display.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I was going to say "well that is what the long date time format for that locale IS" and cannot be changed - OTOH since I've captured it and stored it in `(QString) dlgConnectionProfiles::mDateTimeFormat` I might be able to filter out any appearance of `'t'` (or for safeties sake better include `" t"` as well) for a trailing TZ - just hope there isn't a locale that Qt thinks uses `"(t)"`...

vadi2 commented on src/mudlet.cpp:
```diff
@@ -3257,6 +3257,12 @@ void mudlet::readEarlySettings(const QSettings& settings)
     mInterfaceLanguage = settings.value("interfaceLanguage", autodetectPreferredLanguage()).toString();
+    mUserLocale = QLocale(mInterfaceLanguage);
+    if (mUserLocale == QLocale::c()) {
+        qWarning().nospace().noquote() << "mudlet::readEarlySettings(...) WARNING - Unable to convert language code \"" << mInterfaceLanguage << "\" to a recognised locale, reverting to the POSIX 'C' one...";
```
The ellipsis isn't necessary in any of the statements here.

Kebap replied:
Not sure I understand the suggested change

vadi2 replied:
At the end of the message, the `...` is redundant (or incorrect). `...` is used if there is an action that is going to follow.

SlySven replied:
Well you know how I am with the use of ellipses... :wink: 

... in this case I suppose I was sort of saying "Well that is what has happened, if it breaks then you get to keep the pieces!" But I can do an in-place edit here to trim them out.

Kebap commented:
@SlySven

This has nothing to do with the chosen locale though. Also on your screenshots in discord, there is a time zone shown (as abbreviation, not long form) which is not present in Mudlet release version. 
This change is unexpected and not really required. 

In fact, in some locales it seems to break layout as now the text is suddenly too long for the dropdown and '...' are inserted and hiding the much more important info on day and time of the backups!

I implore you to reconsider.

Kebap commented:
I see there was already a commit addressing the time zone, I will need to evaluate 👍

SlySven commented:
> This has nothing to do with the chosen locale though.

It does because Qt *has* included it in the locale specific date-time format string that it returns when you *ask* it for the one for the particular locale.

As you can see I have now tried to remove it if present - hopefully it won't break things as I have also anticipated the case where the format might have it at the end separated from the preceding elements by a space so that both of them are taken out. It is possible that it might also be presented surrounded by `(`...`)`s - if that is the case then there will be an empty pair of those left - but let us not worry about that case unless it shows up!

> In fact, in some locales it seems to break layout as now the text is suddenly too long for the dropdown and '...' are inserted and hiding the much more important info on day and time of the backups!

That is a significant concern - I think I read somewhere that widgets that are supposed to support the full range of possible locales are supposed to be sized to accommodate text strings that are 150% (i.e. a half as long again) as they need to be for the corresponding *English* texts. This is a case where we may need to increase the (fixed IIRC) width of the right hand side of the dialogue. However that can be left to a separate PR IMHO.

I will wait for a positive response from you @Kebap before I merge this one.

SlySven commented on src/mudlet.cpp:
```diff
@@ -3257,6 +3257,12 @@ void mudlet::readEarlySettings(const QSettings& settings)
     mInterfaceLanguage = settings.value("interfaceLanguage", autodetectPreferredLanguage()).toString();
+    mUserLocale = QLocale(mInterfaceLanguage);
+    if (mUserLocale == QLocale::c()) {
+        qWarning().nospace().noquote() << "mudlet::readEarlySettings(...) WARNING - Unable to convert language code \"" << mInterfaceLanguage << "\" to a recognised locale, reverting to the POSIX 'C' one...";
```
```suggestion
        qWarning().nospace().noquote() << "mudlet::readEarlySettings(...) WARNING - Unable to convert language code \"" << mInterfaceLanguage << "\" to a recognised locale, reverting to the POSIX 'C' one.";
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -3257,6 +3257,12 @@ void mudlet::readEarlySettings(const QSettings& settings)
+    if (mUserLocale == QLocale::c()) {
+        qWarning().nospace().noquote() << "mudlet::readEarlySettings(...) WARNING - Unable to convert language code \"" << mInterfaceLanguage << "\" to a recognised locale, reverting to the POSIX 'C' one...";
+    } else {
+        qDebug().nospace().noquote() << "mudlet::readEarlySettings(...) INFO - Using language code \"" << mInterfaceLanguage << "\" to switch to \"" << QLocale::languageToString(mUserLocale.language()) << " (" << QLocale::countryToString(mUserLocale.country()) << ")\" locale...";
```
```suggestion
        qDebug().nospace().noquote() << "mudlet::readEarlySettings(...) INFO - Using language code \"" << mInterfaceLanguage << "\" to switch to \"" << QLocale::languageToString(mUserLocale.language()) << " (" << QLocale::countryToString(mUserLocale.country()) << ")\" locale.";
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -5415,6 +5421,14 @@ void mudlet::setInterfaceLanguage(const QString& languageCode)
+        mUserLocale = QLocale(mInterfaceLanguage);
+        if (mUserLocale == QLocale::c()) {
+            qWarning().nospace().noquote() << "mudlet::setInterfaceLanguage(\"" << languageCode
+                                           << "\") WARNING - Unable to convert given language code to a recognised locale, reverting to the POSIX 'C' one...";
```
```suggestion
                                           << "\") WARNING - Unable to convert given language code to a recognised locale, reverting to the POSIX 'C' one.";
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -5415,6 +5421,14 @@ void mudlet::setInterfaceLanguage(const QString& languageCode)
+                                           << "\") WARNING - Unable to convert given language code to a recognised locale, reverting to the POSIX 'C' one...";
+        } else {
+            qDebug().nospace().noquote() << "mudlet::setInterfaceLanguage(\"" << languageCode
+                                         << "\") INFO - switching to \"" << QLocale::languageToString(mUserLocale.language()) << " (" << QLocale::countryToString(mUserLocale.country()) << ")\" locale...";
```
```suggestion
                                         << "\") INFO - switching to \"" << QLocale::languageToString(mUserLocale.language()) << " (" << QLocale::countryToString(mUserLocale.country()) << ")\" locale.";
```

SlySven committed:
Apply suggestions from code review

Removed unwanted ellipses from debug texts...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> I think I read somewhere that widgets that are supposed to support the full range of possible locales are supposed to be sized to accommodate text strings that are 150% (i.e. a half as long again) as they need to be for the corresponding _English_ texts. 

https://www.w3.org/International/articles/article-text-size.en

Easily up to 300% especially for brief snippets! English is good at those, others are not always quite.

Enlarging the right-hand side of the connection window seems like a no-go to me, hence my request to remove your newly added time zone info.

vadi2 commented:
I'll leave this for @Kebap to approve on behalf of @Mudlet/ui-review as he's the one who raised this initially.

SlySven committed:
Merge branch 'development' into BugFix_makeDateDisplayOfConnectionsHistoryFollowSelectedUILocale

SlySven commented:
@Kebap please can you indicate if you have any problems with this PR now - without the TZ indication?

Kebap commented:
Didn't get around to look at the new version, yet.

Kebap commented:
Not sure if this merits any more blockage, but just to be complete, I found another locale which may need future fixing:

Spanish with the extra parenthesis () at the end:
![image](https://user-images.githubusercontent.com/117238/86259537-6f134100-bbbc-11ea-8dff-c2c69252111f.png)

vadi2 commented:
Agree should be fixed, we can do this in both this and release-4.10 branches. Merging for now.
